pallets/data-preservers/src/lib.rs

```diff
@@ -62,36 +61,243 @@ use {
+        <T as frame_system::Config>::AccountId,
+    >>::AssignmentWitness;
+
+    /// Data preserver profile.
+    #[derive(
+        RuntimeDebugNoBound,
+        PartialEqNoBound,
+        EqNoBound,
+        Encode,
+        Decode,
+        CloneNoBound,
+        TypeInfo,
+        Serialize,
+        Deserialize,
+    )]
+    #[scale_info(skip_type_params(T))]
+    pub struct Profile<T: Config> {
+        pub url: BoundedVec<u8, T::MaxNodeUrlLen>,
+        pub para_ids: ParaIdsFilter<T>,
+        pub mode: ProfileMode,
+        pub assignment_request: ProviderRequestOf<T>,
+    }
+
+    #[derive(
+        RuntimeDebugNoBound,
+        PartialEqNoBound,
+        EqNoBound,
+        Encode,
+        Decode,
+        CloneNoBound,
+        TypeInfo,
+        Serialize,
+        Deserialize,
+    )]
+    #[scale_info(skip_type_params(T))]
+    pub enum ParaIdsFilter<T: Config> {
+        AnyParaId,
+        Whitelist(BoundedBTreeSet<ParaId, T::MaxParaIdsVecLen>),
+        Blacklist(BoundedBTreeSet<ParaId, T::MaxParaIdsVecLen>),
+    }
+
+    impl<T: Config> ParaIdsFilter<T> {
+        #[allow(clippy::len_without_is_empty)]
+        pub fn len(&self) -> usize {
+            match self {
+                Self::AnyParaId => 0,
+                Self::Whitelist(list) | Self::Blacklist(list) => list.len(),
+            }
+        }
+
+        pub fn can_assign(&self, para_id: &ParaId) -> bool {
+            match self {
+                ParaIdsFilter::AnyParaId => true,
+                ParaIdsFilter::Whitelist(list) => list.contains(para_id),
+                ParaIdsFilter::Blacklist(list) => !list.contains(para_id),
+            }
+        }
+    }
+
+    #[derive(
+        RuntimeDebug, PartialEq, Eq, Encode, Decode, Clone, TypeInfo, Serialize, Deserialize,
+    )]
+    pub enum ProfileMode {
+        Bootnode,
+        Rpc { supports_ethereum_rpcs: bool },
+    }
+
+    /// Profile with additional data:
+    /// - the account id which created (and manage) the profile
+    /// - the amount deposited to register the profile
+    #[derive(
+        RuntimeDebugNoBound,
+        PartialEqNoBound,
+        EqNoBound,
+        Encode,
+        Decode,
+        CloneNoBound,
+        TypeInfo,
+        Serialize,
+        Deserialize,
+    )]
+    #[scale_info(skip_type_params(T))]
+    pub struct RegisteredProfile<T: Config> {
+        pub account: T::AccountId,
+        pub deposit: BalanceOf<T>,
+        pub profile: Profile<T>,
+        /// There can be at most 1 assignment per profile.
+        pub assignment: Option<(ParaId, AssignmentWitnessOf<T>)>,
+    }
+
+    /// Computes the deposit cost of a profile.
+    pub trait ProfileDeposit<Profile, Balance> {
+        fn profile_deposit(profile: &Profile) -> Result<Balance, DispatchErrorWithPostInfo>;
+    }
+
+    /// Implementation of `ProfileDeposit` based on the size of the SCALE-encoding.
+    pub struct BytesProfileDeposit<BaseCost, ByteCost>(PhantomData<(BaseCost, ByteCost)>);
```

Comment:
This is fairly generic ( you only have Profile that requires an encode) I am surprised substrate does not provide anyhting like this alreay. But in any case perhaps this does not belong here but rather somewhere more generic, so that it can be re-used by other pallets. What do you think?

Reply:
I found that substrate has something similar: https://github.com/paritytech/polkadot-sdk/blob/b65313e81465dd730e48d4ce00deb76922618375/substrate/frame/support/src/traits/storage.rs#L183

Essentially you can define a FootPrint structure with the `count` and `len` of profile

Reply:
We could even consider using `HoldConsideration` https://github.com/paritytech/polkadot-sdk/blob/b65313e81465dd730e48d4ce00deb76922618375/substrate/frame/support/src/traits/tokens/fungible/mod.rs#L245